certbot-dns-google/certbot_dns_google/dns_google.py

```diff
@@ -0,0 +1,183 @@
+                  'information about creating a service account and {1} for information about the' +
+                  'required permissions.)').format(ACCT_URL, PERMISSIONS_URL))
+
+    def more_info(self): # pylint: disable=missing-docstring,no-self-use
+        return 'This plugin configures a DNS TXT record to respond to a dns-01 challenge using ' + \
+               'the Google Cloud DNS API.'
+
+    def _setup_credentials(self):
+        self._configure_file('credentials', 'path to Google Cloud DNS service account JSON file')
+
+        dns_common.validate_file_permissions(self.conf('credentials'))
+
+    def _perform(self, domain, validation_name, validation):
+        self._get_google_client().add_txt_record(domain, validation_name, validation, self.ttl)
+
+    def _cleanup(self, domain, validation_name, validation):
+        self._get_google_client().del_txt_record(domain, validation_name, validation, self.ttl)
+
+    def _get_google_client(self):
+        return _GoogleClient(self.conf('credentials'))
+
+
+class _GoogleClient(object):
+    """
+    Encapsulates all communication with the Google Cloud DNS API.
+    """
+
+    def __init__(self, account_json):
+
+        scopes = ['https://www.googleapis.com/auth/ndev.clouddns.readwrite']
+        credentials = ServiceAccountCredentials.from_json_keyfile_name(account_json, scopes)
+        self.dns = discovery.build('dns', 'v1', credentials=credentials, cache_discovery=False)
+        self.project_id = json.load(open(account_json))['project_id']
+
+    def add_txt_record(self, domain, record_name, record_content, record_ttl):
+        """
+        Add a TXT record using the supplied information.
+
+        :param str domain: The domain to use to look up the managed zone.
+        :param str record_name: The record name (typically beginning with '_acme-challenge.').
+        :param str record_content: The record content (typically the challenge validation).
+        :param int record_ttl: The record TTL (number of seconds that the record may be cached).
+        :raises certbot.errors.PluginError: if an error occurs communicating with the Google API
+        """
+
+        zone_id = self._find_managed_zone_id(domain)
+
+        data = {
+            "kind": "dns#change",
+            "additions": [
+                {
+                    "kind": "dns#resourceRecordSet",
+                    "type": "TXT",
+                    "name": record_name + ".",
+                    "rrdatas": [record_content, ],
+                    "ttl": record_ttl,
+                },
+            ],
+        }
+
+        changes = self.dns.changes()  # changes | pylint: disable=no-member
+
+        try:
+            request = changes.create(project=self.project_id, managedZone=zone_id, body=data)
+            response = request.execute()
+
+            status = response['status']
+            change = response['id']
+            while status == 'pending':
```

Comment:
In the Google DNS API, record changes are not synchronous. Once `status` transitions from `pending` to `done`, it has begun propagating.

>  Why do we loop here but not in `del_txt_record` or `_find_managed_zone_id`?

In the case of `del_txt_record`: it doesn't matter if we report cleanup before the change has been made

In the case of `_find_managed_zone_id`: it's a read operation, so there's no status.

>  What is the benefit of calling `request.execute()` repeatedly?

We re-execute the request to get the updated `status` value from the server.